Add KernelLinearOperator, deprecate KeOpsLinearOperator

gpleiss opened the pull request:
KeOpsLinearOperator does not correctly backpropagate gradients if the covar_func closes over parameters.

KernelLinearOperator corrects for this, and is set up to replace LazyEvaluatedKernelTensor in GPyTorch down the line.

[Addresses issues in [#2296](https://github.com/cornellius-gp/gpytorch/pull/2296#issuecomment-1487613562)]

gpleiss committed:
Add KernelLinearOperator, deprecate KeOpsLinearOperator

KeOpsLinearOperator does not correctly backpropagate gradients if the
covar_func closes over parameters.

KernelLinearOperator corrects for this, and is set up to replace
LazyEvaluatedKernelTensor in GPyTorch down the line.

gpleiss commented:
@jacobrgardner @Balandat can one of you take a look at this?

m-julian commented:
This implementation works perfectly when using one KeOps kernel, but there are a few problems when combining kernels. This is not specifically a problem with `KernelLinearOperator`, but the way the things are implemented in `gpytorch`/`linear_operator` currently. Hopefully some of these comments are useful for structuring the code.

I tried to multiply two keops kernels (which make a `ProductKernel`). These are the problems I found:
1. `ProductKernel` calls `to_dense` if x1 and x2 are not identical [here](https://github.com/cornellius-gp/gpytorch/blob/c6a4f46838c26e33795cfc30b6e1fd31d72e26e4/gpytorch/kernels/kernel.py#L653-L658) and returns a `DenseLinearOperator` since `MulLinearOperator` is only used for square symmetric matrices currently. Since KeOps can be used for large matrices where x1 is not equal to x2, calling `to_dense` could give memory errors. Instead, this should give back some object that represents the multiplication of the two symbolic matrices without actually computing them.
2. If x1 is equal to x2, then you end up with a `MulLinearOpeartor` which calls `root_decomposition` on both the `KerneLinearOperator`s. Then there are two issues:
2.1. You might run out of memory if the matrix is very large when doing the root decomposition (because of KeOps).
2.2. If the matrix fits into memory, it will be approximated as Lanczos is the default algorithm. This is fixed by doing `gpytorch.settings.fast_computations(covar_root_decomposition=False)`, but it took me a while to figure out why I was getting different results when using KeOps kernels. If using a very large KeOps symbolic matrix, then I don't think root decomposition should be called, but instead the symbolic matrix should be used until some reduction operation is performed. (So should  `MulLinearOperator` be used with `KernelLinearOperator` instances?)

I was thinking could `KernelLinearOperator` subclass from `DenseLinearOperator` since the `KernelLinearOperator` class is used to represent a full covariance matrix (just computed on the fly when a reduction operation is applied)? That way all the checks for `isinstance(self, DenseLinearOperator)` (for example [this one](https://github.com/cornellius-gp/linear_operator/blob/e69133d2aefd9a5276147bb95588ed6c56a637db/linear_operator/operators/_linear_operator.py#L584-L589)) also work for `KernelLinearOperator`. Then this goes around the problem with `MulLinearOperator` but might cause other problems, so not sure if it is reasonable.

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -350,18 +350,19 @@ def _bilinear_derivative(self, left_vecs: Tensor, right_vecs: Tensor) -> Tuple[O
+        args = []
+        for arg in self.representation():
+            if torch.is_tensor(arg) and arg.dtype.is_floating_point:
+                args.append(arg.detach().requires_grad_(True))
```
Should `requires_grad` be set here depending on whether original `arg` has `requires_grad` set (as was the case before)?

gpleiss replied:
Fixed.

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,217 @@
+
+    .. warning ::
+
+        `covar_func` should not close over any parameters. Any parameters that are closed over will not have
+        propagated gradients.
```
Will everyone who is working on GPs but may not have any formal CS training know what it means to "close over"? Might be worth spelling that out and given an example.

gpleiss replied:
Fixed. I referenced the example in the documentation.

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,217 @@
+    :param covar_func: The covariance function :math:`k_{\boldsymbol \theta}(\cdot, \cdot)`.
+        Its arguments should be `x1`, `x2`, `*params`, `**kwargs`, and it should output the covariance matrix
+        between :math:`\boldsymbol X_1` and :math:`\boldsymbol X_2`.
+    :param kwargs: Any additional (non-hyperparameter) kwargs to pass into `covar_func`.
```
It's a bit weird that all hypers are assumed to be positional args here. It seems that this is rather opaque gotcha, and could also result in making the code much less readable.

gpleiss replied:
@Balandat I can try to allow for kwarg hypers, but this might be a can of worms.

gpleiss replied:
@Balandat okay I changed it so that hypers must be kwargs (cannot be positional args).

This change required modifying some base LinearOperator functionality so that the flattening/unflattening that we do in representation trees (for CG backward passes, etc.) accounts for LinearOperator kwargs as well as args.

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,217 @@
+                        f"Got batch index {batch_indices} but my shape was {self.shape}"
+                    )
+                x2 = self.x2.expand(*([1] * len(batch_indices)), *self.x2.shape[-2:])
+                x2 = x2[(*batch_indices, row_index, dim_index)]
```
you'll probably want to throw this into a helper function to deduplicate the code here

gpleiss replied:
Done.

Balandat commented on test/operators/test_identity_linear_operator.py:
```diff
@@ -76,6 +76,10 @@ def create_linear_op(self):
         return torch.eye(5)
 
+    def test_bilinear_derivative(self):
+        # Not needed, and it errors for some reason
```
Do we know the reason? Might be good to understand

gpleiss replied:
I fixed an issue in LinearOperator._bilinear_derivative (it now exits early if no arguments require gradients), so this short-circuiting is no longer necessary.

gpleiss commented:
@m-julian there is unfortunately no way to do a symbolic element-wise multiplication of two kernels. KeOps (and LinearOperator) can keep things symbolic by using matrix multiplication based algorithms (CG/Lanczos) for solves and log determinants. Unfortunately, matrix multiplication does not distribute over element wise product. The current Lanczos solution comes out of the Product Kernel Interpolation paper (it was the best solution we could come up with).

Therefore, I don't know if there's a better way to handle the product of two matrices than what we currently do in code.

> I was thinking could KernelLinearOperator subclass from DenseLinearOperator since the KernelLinearOperator class is used to represent a full covariance matrix (just computed on the fly when a reduction operation is applied)?

This would probably be a can of worms. DenseLinearOperator assumes that the matrix is represented by a tensor. 
And most of our LinearOperator classes represent full covariance matrices that are computed on the fly when reduction operations are called.

gpleiss committed:
Fix KeOpsLinearOperator deprecation

gpleiss committed:
Allow for kernels with reduced batches and multiple outputs per input

gpleiss committed:
LinearOperator kwargs can be differentiated through

Previously, only positional args were added to the LinearOperator
representation, and so only positional args would receive gradients from
_bilinear_derivative.

This commit also adds Tensor/LinearOperator kwargs to the
representation, and so kwarg Tensor/LinearOperators will also receive
gradients.

gpleiss committed:
Hyperparameters for KernelLinearOperator must be kwargs

gpleiss committed:
LO._bilinear_derivative only computes derivatives for args that require gradients

gpleiss committed:
Expand upon closure variables warning for KernelLinearOperator

gpleiss committed:
LO._bilinear_derivative exits early if no parameters require gradients

gpleiss committed:
Refactor KernelLinearOperator._getitem

Turakar commented:
Regarding the product discussion: Regarding a general implementation that would be part of GPyTorch, I also do not know of a better approach than what @gpleiss pointed out. However, from a user's perspective, you, @m-julian, could of course write a custom KeOps kernel. Most likely, this would even be faster than two separate kernels, as you only need one trip from global memory to processor registers on the GPU.

gpleiss committed:
Allow for optional number of nonbatch parameter dimensions

gpleiss committed:
Fix LO._bilinear_derivative

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -350,17 +359,23 @@ def _bilinear_derivative(self, left_vecs: Tensor, right_vecs: Tensor) -> Tuple[O
+            if torch.is_tensor(arg) and arg.dtype.is_floating_point and arg.requires_grad:
+                args.append(arg.detach().requires_grad_(True))
+            else:
+                args.append(arg.detach())
```
may want to comment here that the representation is guaranteed to consist only or tensors

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -150,7 +150,16 @@ def __init__(self, *args, **kwargs):
+            # Sorting is necessary so that the flattening in the representation tree is deterministic
+            if torch.is_tensor(val) or isinstance(val, LinearOperator):
+                self._differentiable_kwarg_names.append(name)
+                self._differentiable_kwarg_vals.append(val)
```
Why not just use an (ordered) dict for this?

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -457,6 +472,16 @@ def _args(self) -> Tuple[Union[torch.Tensor, "LinearOperator", int], ...]:
         self._args_memo = args
 
+    @property
+    def _differentiable_kwargs(self) -> Dict[str, Union[Tensor, "LinearOperator"]]:
```
could avoid this by using an ordered dict to store these it seems?

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -457,6 +472,16 @@ def _args(self) -> Tuple[Union[torch.Tensor, "LinearOperator", int], ...]:
+    @property
+    def _kwargs(self) -> Dict[str, Any]:
+        kwargs = self._differentiable_kwargs
+        kwargs.update(self._nondifferentiable_kwargs)
```
hmm this only works b/c you're generating the `_differentiable_kwargs` dynamically as a property. Worth commenting on this here. Alternatively just use a dict to store these, and then do sth like 

    @property
    def _kwargs(self) -> Dict[str, Any]:
        return {**self._differentiable_kwargs, **self._nondifferentiable_kwargs}

here.

Balandat commented on linear_operator/operators/_linear_operator.py:
```diff
@@ -2013,7 +2042,7 @@ def representation(self) -> Tuple[torch.Tensor, ...]:
         """
         representation = []
-        for arg in self._args:
+        for arg in list(self._args) + list(self._differentiable_kwarg_vals):
```
Can be agnostic to the types here if you just want to iterate
```suggestion
        for arg in itertools.chain(self._args, self._differentiable_kwarg_vals):
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+        if isinstance(batch_indices, slice):
+            x = x.expand(1, *x.shape[-2:])[(*batch_indices, data_index, _noop_index)]
+        elif isinstance(batch_indices, tuple):
+            if any([not isinstance(bi, slice) for bi in batch_indices]):
```
```suggestion
            if any(not isinstance(bi, slice) for bi in batch_indices):
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+    # This way - in the default case, we can avoid doing expansions of x1 which can be timely
+    except IndexError:
+        if isinstance(batch_indices, slice):
+            x = x.expand(1, *x.shape[-2:])[(*batch_indices, data_index, _noop_index)]
```
Readability 
```suggestion
            x = x.expand(1, *x.shape[-2:])
            x = [(*batch_indices, data_index, _noop_index)]
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+    :param num_outputs_per_input: The number of outputs per data point.
+        This parameter should be 1 for most kernels, but will be >1 for multitask kernels,
+        gradient kernels, and any other kernels that require cross-covariance terms for multiple domains.
+        If a tuple is passed, there will be a different number of outputs per input dimension
+        for the rows/cols of the kernel matrix.
```
Do we need the convenience of just passing an int or should we just require this to be defined explicitly?

gpleiss replied:
We can define it explicitly.

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+        x2: Float[Tensor, "... N D"],
+        covar_func: Callable[..., Float[Union[Tensor, LinearOperator], "... M N"]],
+        num_outputs_per_input: Union[int, Tuple[int, int]] = (1, 1),
+        num_nonbatch_dimensions: Dict[str, int] = dict(),
```
linters are going to hate creating a dict in the function signature. Suggest using `Optional[Dict[str, int]]= None` instead

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+            else:
+                param_batch_shapes[name] = val.shape[: -num_nonbatch_dimensions[name]]
+                param_nonbatch_shapes[name] = val.shape[-num_nonbatch_dimensions[name] :]
+
```
```suggestion
                nonbatch_dim = num_nonbatch_dimensions[name]
                param_batch_shapes[name] = val.shape[: -nonbatch_dim]
                param_nonbatch_shapes[name] = val.shape[-nonbatch_dim :]

```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+                    val.expand(*batch_broadcast_shape, *param_nonbatch_shapes[name]).requires_grad_(val.requires_grad),
+                )
+                for name, val in tensor_params.items()
+            )
```
```suggestion
            tensor_params = {
                name : val.expand(
                    *batch_broadcast_shape, *param_nonbatch_shapes[name]
                ).requires_grad_(val.requires_grad)
                for name, val in tensor_params.items()
            }
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+        # and then squeeze out the batch dimensions
+        x1 = self.x1.unsqueeze(0).transpose(0, -2)
+        x2 = self.x2.unsqueeze(0).transpose(0, -2)
+        tensor_params = dict((name, val.unsqueeze(0)) for name, val in self.tensor_params.items())
```
```suggestion
        tensor_params = {name: val.unsqueeze(0) for name, val in self.tensor_params.items()}
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+    def _get_indices(self, row_index: IndexType, col_index: IndexType, *batch_indices: IndexType) -> torch.Tensor:
+        x1_ = self.x1[(*batch_indices, row_index)].unsqueeze(-2)
+        x2_ = self.x2[(*batch_indices, col_index)].unsqueeze(-2)
+        tensor_params_ = dict((name, val[batch_indices]) for name, val in self.tensor_params.items())
```
```suggestion
        tensor_params_ = {name: val[batch_indices] for name, val in self.tensor_params.items()}
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+        tensor_params = dict(
+            (name, val[(*batch_indices, *([_noop_index] * self.num_nonbatch_dimensions[name]))])
+            for name, val in self.tensor_params.items()
+        )
```
```suggestion
        tensor_params = {
            name: val[(*batch_indices, *([_noop_index] * self.num_nonbatch_dimensions[name]))]
            for name, val in self.tensor_params.items()
        }
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+        tensor_params = dict(
+            (name, val.permute(*dims, *range(-self.num_nonbatch_dimensions[name], 0)))
+            for name, val in self.tensor_params.items()
+        )
```
```suggestion
        tensor_params = {
            name: val.permute(*dims, *range(-self.num_nonbatch_dimensions[name], 0))
            for name, val in self.tensor_params.items()
        }
```

Balandat commented on linear_operator/operators/kernel_linear_operator.py:
```diff
@@ -0,0 +1,385 @@
+    def _unsqueeze_batch(self, dim: int) -> LinearOperator:
+        x1 = self.x1.unsqueeze(dim)
+        x2 = self.x2.unsqueeze(dim)
+        tensor_params = dict((name, val.unsqueeze(dim)) for name, val in self.tensor_params.items())
```
```suggestion
        tensor_params = {name: val.unsqueeze(dim) for name, val in self.tensor_params.items()}
```

Balandat commented on linear_operator/operators/linear_operator_representation_tree.py:
```diff
@@ -4,11 +4,12 @@
         counter = 0
         self.children = []
-        for arg in linear_op._args:
+        for arg in list(linear_op._args) + list(linear_op._differentiable_kwarg_vals):
```
```suggestion
        for arg in itertools.chain(linear_op._args, linear_op._differentiable_kwarg_vals):
```

Balandat reviewed (approved):
I can't say I did check all the indexing logic in exhaustive detail, but hopefully we have some test coverage for that?

gpleiss committed:
Update linear_operator/operators/_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/linear_operator_representation_tree.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Update linear_operator/operators/kernel_linear_operator.py

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

gpleiss committed:
Fix errors, address comments

gpleiss committed:
KroneckerProductLinearOperator broadcasts

gpleiss committed:
Test cases and fixes for multitask KernelLinearOperator

Balandat commented on linear_operator/operators/kronecker_product_linear_operator.py:
```diff
@@ -62,23 +62,36 @@ def _t_matmul(linear_ops, kp_shape, rhs):
+                "are incompatible for a Kronecker product."
+            )
+
+        if len(batch_broadcast_shape):  # Otherwise all linear_ops are non-batch, and we don't need to expand
```
This introduced an inconsistency in the type of `linear_ops` (a `list` rather than a `tuple`), which resulted in some downstream breakages in botorch. Fixed in #66.
